hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,471 @@
+    y_test,
+    y_pred,
+    y_train=None,
+    ipcw_estimator="km",
+    time_grid=None,
+    taus=None,
+    event_of_interest=1,
+    tied_tol=1e-8,
+):
+    r"""Time-dependent concordance index for prognostic models with competing risks \
+    using inverse probability of censoring weighting.
+
+    .. math::
+
+        \mathrm{C}(t) = \frac{\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) Q^{ij}(t)
+        \tilde{N}^1_i(t)}
+        {\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) \tilde{N}^1_i(t)}
+
+    where:
+
+    .. math::
+
+        \begin{align}
+        \tilde{N}^1_i(t) &= I\{\tilde{T}_i \leq t, \tilde{D}_i = 1\} \\
+        \tilde{A}_{ij} &= I\{\tilde{T}_i < \tilde{T}_j \cup (\tilde{T}_i =
+        \tilde{T}_j \cap D_j = 0)\} \\
+        \tilde{B}_{ij} &= I\{\tilde{T}_i \geq \tilde{T}_j, D_j = 2\} \\
+        \hat{W}_{ij,1} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_i|X_j) \\
+        \hat{W}_{ij,2} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_j-|X_j) \\
+        Q_{ij}(t) &= I\{M(t, X_i) > M(t, X_j)\}
+        \end{align}
+
+    where :math:`D_j = 0`, :math:`D_j = 1` and :math:`D_j = 2` respectively indicate
+    individuals having been censored, individuals having experienced the event of
+    interest, and individual having experienced a competing event. :math:`\hat{G}`
+    is a IPCW estimator, :math:`Q_{ij}(t)` is an indicator for the order of
+    predicted risk at :math:`t`, and :math:`M` is the predicted cumulative incidence
+    function for the event of interest.
+
+    The concordance index (C-index) is a common metric in survival analysis that
+    evaluates whether the model predictions correctly order pairs of individuals with
+    respect to the timing of the event of interest. It is defined as the probability
+    that the prediction is concordant for a pair of individuals and is computed as the
+    ratio of the number of concordant pairs to the total number of pairs.
+    This implementation extends the C-index to the competing risks setting, where
+    multiple alternative events are considered, aiming to determine which event occurs
+    first and when, following the formulas and notations in [Wolbers2014]_.
+
+    Due to the right-censoring in the data, the order of some pairs is unknown,
+    so we define the notion of comparable pairs, i.e. the pairs for which
+    we can compare the order of occurrence of the event of interest.
+    A pair :math:`(i, j)` is comparable, with :math:`i` experiencing the event of
+    interest at time :math:`T_i` if:
+
+    - :math:`j` is censored or experience any event at a strictly greater time
+      :math:`T_j > T_i` (pair of type A)
+    - :math:`j` is censored at the exact same time :math:`T_i = T_j` (pair of type A).
+    - :math:`j` experiences a competing event before or at time :math:`T_i`
+      (pair of type B)
```

Comment:
the docstring should specify that this is the event-specific concordance index, to explicit why we distinguish the event and the competing events (maybe at the beginning)